src/metrics_clustering.rs

```diff
@@ -0,0 +1,221 @@
+use crate::error::Result;
+use crate::Float;
+use ndarray::{ArrayBase, ArrayView1, Data, Ix2};
+use ndarray_stats::DeviationExt;
+use std::collections::HashMap;
+
+/// Evaluates the quality of a clustering using euclidean distance.
+pub trait SilhouetteScore<'a, F> {
+    /// Evaluates the quality of a clustering.
+    ///
+    /// Given a clustered dataset,
+    /// the silhouette score for each sample is computed as
+    /// the relative difference between the average distance
+    /// of the sample to other samples in the same cluster and
+    /// the minimum average distance of the sample to samples in
+    /// another cluster. This value goes from -1 to +1 when the point
+    /// is respectively closer (in average) to points in another cluster and to points in its own cluster.
+    ///
+    /// Finally, the silhouette score for the clustering is evaluated as the mean
+    /// silhouette score of each sample.
+    fn silhouette_score(&'a self) -> Result<F>;
+}
+
+struct DistanceCount<F> {
+    total_distance: F,
+    count: usize,
+}
+
+impl<F: Float> DistanceCount<F> {
+    /// Sets the total distance from the sample to this cluster to zero
+    pub fn reset(&mut self) {
+        self.total_distance = F::zero();
+    }
+
+    pub fn new(count: usize) -> DistanceCount<F> {
+        DistanceCount {
+            total_distance: F::zero(),
+            count,
+        }
+    }
+
+    /// Divides the total distance from the sample to this cluster by the number of samples in the cluster
+    pub fn mean_distance(&self) -> F {
+        self.total_distance / F::from(self.count).unwrap()
+    }
+
+    /// To be used in the cluster in which the sample is located. The distance from the sample to itself
+    /// is zero so it does not get added to the total distance. We can then just divide the total
+    /// distance by 1 - #samples in this cluster
+    pub fn same_label_mean_distance(&self) -> F {
+        if self.count == 1 {
+            return F::zero();
+        }
+        self.total_distance / F::from(self.count - 1).unwrap()
+    }
+
+    /// adds the distance of `other_sample` from `eval_sample` to the total distance of `eval_sample` from the current cluster
+    pub fn add_point(&mut self, eval_sample: ArrayView1<F>, other_sample: ArrayView1<F>) {
+        self.total_distance += F::from(eval_sample.l2_dist(&other_sample).unwrap()).unwrap();
+    }
+}
+
+impl<
+        'a,
+        F: Float,
+        L: 'a + Label,
+        D: Data<Elem = F>,
+        T: AsTargets<Elem = L> + FromTargetArray<'a, L> + Labels<Elem = L>,
+    > SilhouetteScore<'a, F> for DatasetBase<ArrayBase<D, Ix2>, T>
+{
+    fn silhouette_score(&'a self) -> Result<F> {
+        // By using try_single_target we ensure that the iterator returns an
+        // array1 as target with just one element, that can be addressed by [0]
+        let mut labels: HashMap<L, DistanceCount<F>> =
+            self.targets().try_single_target()?.label_count()[0]
```

Comment:
I see, I would still need to use 
```
self.label_count()
                .into_iter()
                .flatten()
                .into_iter()
                .map(|..| ..)
                .collect();
```
to be able to iterate over the hashmap instead of the vector with only one hashmap in it

Reply:
damn should have looked at the definition more closely :see_no_evil: 
then I would just use `self.label_counter().remove(0).into_iter().map(|..| ..)`